File: public/java/039-kth-smallest-element-in-bst.java
/*
 * LeetCode Problem #230: Kth Smallest Element in a BST
 * Difficulty: Medium
 * Link: https://leetcode.com/problems/kth-smallest-element-in-a-bst/
 */
import java.util.*;

public class Solution {

    static class TreeNode {
        int val;
        TreeNode left, right;
        TreeNode(int val) { this.val = val; }
    }

    // APPROACH 1: Recursive Inorder | O(k) time | O(h) space
    // EXPLAIN: Inorder traversal of a BST yields sorted order; stop when k nodes have been visited.
    private int count, result;

    public int kthSmallest(TreeNode root, int k) {
        count = 0;
        inorder(root, k);
        return result;
    }
    private void inorder(TreeNode node, int k) {
        if (node == null || count >= k) return;
        inorder(node.left, k);
        if (++count == k) { result = node.val; return; }
        inorder(node.right, k);
    }

    // APPROACH 2: Iterative Inorder | O(h + k) time | O(h) space
    // EXPLAIN: Stack-based inorder traversal; pop nodes one by one and return when k-th is reached.
    public int kthSmallestIterative(TreeNode root, int k) {
        Deque<TreeNode> stack = new ArrayDeque<>();
        TreeNode curr = root;
        int cnt = 0;
        while (curr != null || !stack.isEmpty()) {
            while (curr != null) { stack.push(curr); curr = curr.left; }
            curr = stack.pop();
            if (++cnt == k) return curr.val;
            curr = curr.right;
        }
        return -1;
    }

    // APPROACH 3: Store Inorder in Array | O(n) time | O(n) space
    // EXPLAIN: Collect all values via inorder traversal; return the element at index k-1.
    public int kthSmallestArray(TreeNode root, int k) {
        List<Integer> vals = new ArrayList<>();
        collectInorder(root, vals);
        return vals.get(k - 1);
    }
    private void collectInorder(TreeNode node, List<Integer> vals) {
        if (node == null) return;
        collectInorder(node.left, vals);
        vals.add(node.val);
        collectInorder(node.right, vals);
    }

    // APPROACH 4: Morris Traversal | O(n) time | O(1) space
    // EXPLAIN: Threaded-tree inorder; count nodes without extra memory.
    public int kthSmallestMorris(TreeNode root, int k) {
        TreeNode curr = root;
        int cnt = 0;
        while (curr != null) {
            if (curr.left == null) {
                if (++cnt == k) return curr.val;
                curr = curr.right;
            } else {
                TreeNode pred = curr.left;
                while (pred.right != null && pred.right != curr) pred = pred.right;
                if (pred.right == null) {
                    pred.right = curr; curr = curr.left;
                } else {
                    pred.right = null;
                    if (++cnt == k) return curr.val;
                    curr = curr.right;
                }
            }
        }
        return -1;
    }

    // APPROACH 5: Augmented BST (Follow-up) | O(h) time | O(n) space
    // EXPLAIN: Annotate each node with its left-subtree size; binary-search for k in O(h) per query.
    static class AugNode {
        int val, leftSize;
        AugNode left, right;
        AugNode(int v) { val = v; }
    }

    public int kthSmallestAugmented(TreeNode root, int k) {
        AugNode aug = buildAug(root);
        return searchAug(aug, k);
    }
    private int countNodes(AugNode node) {
        if (node == null) return 0;
        return 1 + countNodes(node.left) + countNodes(node.right);
    }
    private AugNode buildAug(TreeNode node) {
        if (node == null) return null;
        AugNode aug = new AugNode(node.val);
        aug.left  = buildAug(node.left);
        aug.right = buildAug(node.right);
        aug.leftSize = countNodes(aug.left);
        return aug;
    }
    private int searchAug(AugNode node, int k) {
        if (k <= node.leftSize)        return searchAug(node.left,  k);
        if (k == node.leftSize + 1)    return node.val;
        return searchAug(node.right, k - node.leftSize - 1);
    }

    // APPROACH 6: Standard Solution | O(h + k) time | O(h) space
    // EXPLAIN: Canonical iterative inorder — same as Approach 2; matches typical LeetCode submission.
    public int kthSmallestStandard(TreeNode root, int k) {
        return kthSmallestIterative(root, k);
    }
}

// Made with Bob

File: public/python/039-kth-smallest-element-in-bst.py
# LeetCode Problem #230: Kth Smallest Element in a BST
# Difficulty: Medium
# Link: https://leetcode.com/problems/kth-smallest-element-in-a-bst/

from typing import Optional, List


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


# ─────────────────────────────────────────────
# APPROACH 1: Recursive Inorder | O(k) time | O(h) space
# EXPLAIN: Inorder traversal of a BST yields sorted order; stop and record when the k-th node is reached.
# WHEN: Clean recursive solution; early exit after k nodes.

def kthSmallest_recursive(root: Optional[TreeNode], k: int) -> int:
    result = [0]
    count = [0]

    def inorder(node):
        if not node or count[0] >= k:
            return
        inorder(node.left)
        count[0] += 1
        if count[0] == k:
            result[0] = node.val
            return
        inorder(node.right)

    inorder(root)
    return result[0]


# ─────────────────────────────────────────────
# APPROACH 2: Iterative Inorder | O(h + k) time | O(h) space
# EXPLAIN: Explicit stack-based inorder; pop and count nodes until the k-th is reached.
# WHEN: Avoids recursion; preferred for large trees or restricted environments.

def kthSmallest_iterative(root: Optional[TreeNode], k: int) -> int:
    stack = []
    curr = root
    count = 0
    while curr or stack:
        while curr:
            stack.append(curr)
            curr = curr.left
        curr = stack.pop()
        count += 1
        if count == k:
            return curr.val
        curr = curr.right
    return -1


# ─────────────────────────────────────────────
# APPROACH 3: Store Inorder in Array | O(n) time | O(n) space
# EXPLAIN: Collect all values in sorted order then return index k-1.
# WHEN: Simplest logic; acceptable when the full traversal cost is fine.

def kthSmallest_array(root: Optional[TreeNode], k: int) -> int:
    values: List[int] = []

    def collect(node):
        if not node:
            return
        collect(node.left)
        values.append(node.val)
        collect(node.right)

    collect(root)
    return values[k - 1]


# ─────────────────────────────────────────────
# APPROACH 4: Morris Traversal | O(n) time | O(1) space
# EXPLAIN: Threaded-tree inorder traversal; count nodes without extra memory.
# WHEN: O(1) extra space required; complex but avoids any auxiliary data structure.

def kthSmallest_morris(root: Optional[TreeNode], k: int) -> int:
    curr = root
    count = 0
    while curr:
        if not curr.left:
            count += 1
            if count == k:
                return curr.val
            curr = curr.right
        else:
            pred = curr.left
            while pred.right and pred.right is not curr:
                pred = pred.right
            if not pred.right:
                pred.right = curr
                curr = curr.left
            else:
                pred.right = None
                count += 1
                if count == k:
                    return curr.val
                curr = curr.right
    return -1


# ─────────────────────────────────────────────
# APPROACH 5: Augmented BST (Follow-up) | O(h) time per query | O(n) space
# EXPLAIN: Each node stores its left-subtree size; binary search to find k-th in O(h) per query.
# WHEN: Optimal for frequent kth queries after preprocessing; O(h) per query vs O(h+k).

class AugmentedNode:
    def __init__(self, val):
        self.val = val
        self.left = self.right = None
        self.left_size = 0  # number of nodes in left subtree

def kthSmallest_augmented(root: Optional[TreeNode], k: int) -> int:
    # Build augmented tree
    def build(node: Optional[TreeNode]) -> Optional[AugmentedNode]:
        if not node:
            return None
        aug = AugmentedNode(node.val)
        aug.left  = build(node.left)
        aug.right = build(node.right)
        aug.left_size = (aug.left.left_size + 1 if aug.left else 0) + (count_nodes(aug.left))
        return aug

    def count_nodes(node) -> int:
        if not node:
            return 0
        return 1 + count_nodes(node.left) + count_nodes(node.right)

    def search(node, k):
        left_count = node.left_size
        if k <= left_count:
            return search(node.left, k)
        elif k == left_count + 1:
            return node.val
        else:
            return search(node.right, k - left_count - 1)

    aug_root = build(root)
    return search(aug_root, k)


# ─────────────────────────────────────────────
# APPROACH 6: Standard Solution (most common) | O(h + k) time | O(h) space
# EXPLAIN: Canonical iterative inorder — same as Approach 2; matches typical interview submission.
# WHEN: Default answer format.

def kthSmallest(root: Optional[TreeNode], k: int) -> int:
    return kthSmallest_iterative(root, k)


# Made with Bob
